### DISCUSSION TITLE
What is the best practice to fetch dynamic data in layout?

### DISCUSSION THREAD

**Discussion by sshvaiko** [vercel/next.js#58095] [App Router]
Hi everyone! We need to fetch some dynamic data (we can do it at build time) in Layout. For example, menu config for the header component.

We tried to do it this way:
```
export default async function RootLayout({ children }: any) {
  const data = await getData();
```

But I have noticed that the website has become super slow. It looked like this code was run whenever a user opened any page. But I would like to run this code only once during the build and then use this menu config on all pages.

What is the best way to achieve it?

**Comment by mutasim77** [comment]
> this code was run whenever a user opened any page

What about storing the data globally and checking if we already have the data, so we don't need to fetch it again?

**Reply by manavm1990** [reply]
Globally like with **context?** That might create some complications regarding **server** and **client** components, but it might be feasible.

**Comment by manavm1990** [comment]
I'm NS 😕 if I quite get your question @sshvaiko. You might need to provide a bit more info ℹ️ such as your component tree 🌴 .

As for my code, I have this:

```tsx
export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const config = await getConfig();
  const user = getCurrentUser();

  return (
    <html lang="en">
      <TwicInstall domain={TWIC_PICS_DOMAIN} />
      <body className={commissioner.className}>
        <Header
          config={config}
```

I use **server component** (`layout.tsx`) to do the fetch, and then pass those **serializable props** to the **client component.**

**Comment by sshvaiko** [comment]
@manavm1990 this is exactly what I need and what I tried. But in this case, it looks like getConfig and getCurrentUser will be run for **each** page that uses this layout.

And it affects performance.

**Reply by manavm1990** [reply]
> f you need to use the same data (e.g. current user) in multiple components in a tree, you do not have to fetch data globally, nor forward props between components. Instead, you can use fetch or React cache in the component that needs the data without worrying about the performance implications of making multiple requests for the same data.

> This is possible because fetch requests are automatically memoized. Learn more about [request memoization](https://nextjs.org/docs/app/building-your-application/caching#request-memoization)
> [Source](https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#fetching-data-where-its-needed)

This implies that repeated fetches of the same data are irrelevant from other **server components.**

I did like this 1️⃣ :

```diff
diff --git a/app/lib/fetchers.ts b/app/lib/fetchers.ts
index 5dc13cf..346499c 100644
--- a/app/lib/fetchers.ts
+++ b/app/lib/fetchers.ts
@@ -2,6 +2,8 @@ import { configSchema, type Config } from '@/types/config.types';
 import { MOCK_API_KEY, MOCK_API_URL } from '@/types/env.types';
 
 export async function getConfig(): Promise<Config> {
+  console.log('Fetching config...');
+
   const response = await fetch(`${MOCK_API_URL}/app/config`, {
     headers: {
       'x-api-key': MOCK_API_KEY,
```

I 'cold started' the app in 'Incognito' mode. I switched from 'home' to another full page and I saw this in the console (`layout` is the same).

I saw 👀:

```shell
✓ Ready in 801ms
 ○ Compiling /page ...
 ✓ Compiled /page in 4.5s
Fetching config...
Fetching config...
 ✓ Compiled /favicon.ico/route in 69ms
 ○ Compiling /contact/page ...
 ✓ Compiled /contact/page in 936ms
 ```
 
 Thre are 2️⃣ `'Fetching config...'` b/c of 'dev mode,' but it didn't 🔥 off when switching to other page.
